File: test/classify_agent_authorship_test.rb
# frozen_string_literal: true

require "json"
require "minitest/autorun"
require "open3"
require "tempfile"

class ClassifyAgentAuthorshipTest < Minitest::Test
  ROOT = File.expand_path("..", __dir__)
  SCRIPT = File.join(ROOT, ".github/scripts/classify-agent-authorship.rb")

  def test_untrailered_commits_are_agent_assisted
    outputs = classify([{ "sha" => "abc", "message" => "fix: regular change" }])

    assert_equal "agent-assisted", outputs.fetch("label")
    assert_equal "1", outputs.fetch("total_commits")
    assert_equal "0", outputs.fetch("agent_commits")
    assert_equal "1", outputs.fetch("untrailered_commits")
    assert_equal "0", outputs.fetch("incomplete_agent_commits")
  end

  def test_complete_maestro_trailers_are_agent_authored
    outputs = classify([{ "sha" => "abc", "message" => <<~MSG }])
      feat: ship change

      Co-Authored-By: Maestro <maestro@evalops.dev>
      Maestro-Version: 2026.04.28 / gpt-5
      Maestro-Prompt-Id: prompt-123
      Maestro-Approvals-Id: approval-456
    MSG

    assert_equal "agent-authored", outputs.fetch("label")
    assert_equal "1", outputs.fetch("agent_commits")
    assert_equal "0", outputs.fetch("untrailered_commits")
    assert_equal "0", outputs.fetch("incomplete_agent_commits")
  end

  def test_mixed_authorship_and_incomplete_trailers_are_reported
    outputs = classify(
      [
        { "sha" => "abc", "message" => <<~MSG },
          feat: partial agent change

          Co-Authored-By: Maestro <maestro@evalops.dev>
          Maestro-Version: 2026.04.28 / gpt-5
        MSG
        { "sha" => "def", "message" => "docs: human follow-up" },
      ],
    )

    assert_equal "mixed-authorship", outputs.fetch("label")
    assert_equal "1", outputs.fetch("agent_commits")
    assert_equal "1", outputs.fetch("untrailered_commits")
    assert_equal "1", outputs.fetch("incomplete_agent_commits")
  end

  def test_github_output_file_gets_same_outputs
    Tempfile.create("github-output") do |file|
      outputs = classify(
        [{ "sha" => "abc", "message" => "fix: regular change" }],
        github_output: file.path,
      )
      file_outputs = parse_outputs(File.read(file.path))

      assert_equal outputs, file_outputs
    end
  end

  private

  def classify(commits, github_output: nil)
    input = commits.map(&:to_json).join("\n")
    args = ["ruby", SCRIPT]
    args += ["--github-output", github_output] if github_output
    stdout, stderr, status = Open3.capture3(*args, stdin_data: input)
    assert status.success?, stderr
    parse_outputs(stdout)
  end

  def parse_outputs(text)
    text.each_line(chomp: true).to_h { |line| line.split("=", 2) }
  end
end

File: test/classify_agent_authorship_test.py
import json
import subprocess
import sys
import tempfile
import unittest
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SCRIPT = ROOT / ".github" / "scripts" / "classify_agent_authorship.py"


def parse_outputs(text):
    return dict(line.split("=", 1) for line in text.splitlines())


def classify(commits, github_output=None):
    input_data = "\n".join(json.dumps(commit) for commit in commits)
    args = [sys.executable, str(SCRIPT)]
    if github_output:
        args += ["--github-output", github_output]
    result = subprocess.run(
        args, input=input_data, capture_output=True, text=True
    )
    assert result.returncode == 0, result.stderr
    return parse_outputs(result.stdout)


class ClassifyAgentAuthorshipTest(unittest.TestCase):
    def test_untrailered_commits_are_agent_assisted(self):
        outputs = classify([{"sha": "abc", "message": "fix: regular change"}])

        self.assertEqual("agent-assisted", outputs["label"])
        self.assertEqual("1", outputs["total_commits"])
        self.assertEqual("0", outputs["agent_commits"])
        self.assertEqual("1", outputs["untrailered_commits"])
        self.assertEqual("0", outputs["incomplete_agent_commits"])

    def test_complete_maestro_trailers_are_agent_authored(self):
        outputs = classify(
            [
                {
                    "sha": "abc",
                    "message": (
                        "feat: ship change\n"
                        "\n"
                        "Co-Authored-By: Maestro <maestro@evalops.dev>\n"
                        "Maestro-Version: 2026.04.28 / gpt-5\n"
                        "Maestro-Prompt-Id: prompt-123\n"
                        "Maestro-Approvals-Id: approval-456\n"
                    ),
                }
            ]
        )

        self.assertEqual("agent-authored", outputs["label"])
        self.assertEqual("1", outputs["agent_commits"])
        self.assertEqual("0", outputs["untrailered_commits"])
        self.assertEqual("0", outputs["incomplete_agent_commits"])

    def test_mixed_authorship_and_incomplete_trailers_are_reported(self):
        outputs = classify(
            [
                {
                    "sha": "abc",
                    "message": (
                        "feat: partial agent change\n"
                        "\n"
                        "Co-Authored-By: Maestro <maestro@evalops.dev>\n"
                        "Maestro-Version: 2026.04.28 / gpt-5\n"
                    ),
                },
                {"sha": "def", "message": "docs: human follow-up"},
            ]
        )

        self.assertEqual("mixed-authorship", outputs["label"])
        self.assertEqual("1", outputs["agent_commits"])
        self.assertEqual("1", outputs["untrailered_commits"])
        self.assertEqual("1", outputs["incomplete_agent_commits"])

    def test_github_output_file_gets_same_outputs(self):
        with tempfile.NamedTemporaryFile(mode="r", suffix="github-output") as handle:
            outputs = classify(
                [{"sha": "abc", "message": "fix: regular change"}],
                github_output=handle.name,
            )
            file_outputs = parse_outputs(handle.read())

        self.assertEqual(outputs, file_outputs)


if __name__ == "__main__":
    unittest.main()
